repose-aggregator/components/filters/body-patcher-filter/src/main/scala/org/openrepose/filters/bodypatcher/BodyPatcherFilter.scala

```diff
@@ -0,0 +1,150 @@
+  extends AbstractConfiguredFilter[BodyPatcherConfig](configurationService) with LazyLogging {
+  override val DEFAULT_CONFIG: String = "body-patcher.cfg.xml"
+  override val SCHEMA_LOCATION: String = "/META-INF/schema/config/body-patcher.xsd"
+
+  override def doWork(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = {
+    val httpRequest: HttpServletRequest = request.asInstanceOf[HttpServletRequest]
+    val httpResponse: HttpServletResponse = response.asInstanceOf[HttpServletResponse]
+
+    val pathChanges: List[ChangeDetails] = filterChanges(httpRequest)
+    val requestPatches: List[Patch] = filterRequestChanges(pathChanges)
+    val responsePatches: List[Patch] = filterResponseChanges(pathChanges)
+
+    //patch the request
+    val patchedRequest: HttpServletRequest = determineContentType(httpRequest.getContentType) match {
+      case Json =>
+        val jsonPatches: List[String] = filterJsonPatches(requestPatches)
+        if (jsonPatches.nonEmpty) {
+          logger.debug("Applying json patches on the request")
+          val patchedValue: JsValue = applyJsonPatches(PJson.parse(httpRequest.getInputStream), jsonPatches)
+          new HttpServletRequestWrapper(httpRequest, new ServletInputStreamWrapper(new ByteArrayInputStream(PJson.stringify(patchedValue).getBytes)))
+        } else {
+          httpRequest
+        }
+      case Xml =>
+        //todo: implement xml support
+        httpRequest
+      case Other =>
+        httpRequest
+    }
+
+
+    //check if we might do work to the response
+    if (responsePatches.isEmpty) {
+      chain.doFilter(patchedRequest, httpResponse)
+    } else {
+      val wrappedResponse: HttpServletResponseWrapper = new HttpServletResponseWrapper(httpResponse, PASSTHROUGH, MUTABLE)
+      chain.doFilter(patchedRequest, wrappedResponse)
+      wrappedResponse.uncommit()
+
+      determineContentType(wrappedResponse.getContentType) match {
+        case Json =>
+          val jsonPatches: List[String] = filterJsonPatches(responsePatches)
+          if (jsonPatches.nonEmpty) {
+            logger.debug("Applying json patches on the response")
+
+            val contentStream = wrappedResponse.getOutputStreamAsInputStream
+            val contentValue = if (contentStream.available() > 0) {
+              PJson.parse(contentStream)
+            } else {
+              PJson.obj()
+            }
+            val patchedValue: JsValue = applyJsonPatches(contentValue, jsonPatches)
+            wrappedResponse.setOutput(new ByteArrayInputStream(PJson.stringify(patchedValue).getBytes))
+          }
+        case Xml => //todo implement xml support
+        case Other =>
+      }
+      wrappedResponse.commitToResponse()
+    }
+  }
+
+  def filterChanges(request: HttpServletRequest): List[ChangeDetails] = {
+    val urlPath: String = new URL(request.getRequestURL.toString).getPath
+    configuration.getChange.asScala.toList
```

Comment:
Odd. I would have expected it to return a standard type higher up on the hierarchy.

I think technically it returns a wrapper around the Java collection, which costs next to nothing. `toList` may actually re-arrange the data if it's not in a list-like format, though.

Anyway, I'm not worried about it. It will almost certainly be a tiny list anyway.